File: server/backend/src/blockchain_clients/bitcoin_client.py
"""
server/backend/src/blockchain_clients/bitcoin_client.py
=======================================================================
Bitcoin JSON-RPC client for the Multichain Block Explorer.

Reads connection config from environment variables.
Used exclusively by controllers.py.

Usage:
    from .blockchain_clients.bitcoin_client import BitcoinClient

    client = BitcoinClient()
    result = client.call("getblockchaininfo")
"""

import os
import requests
from requests.auth import HTTPBasicAuth
from dotenv import load_dotenv

load_dotenv()


# ---------------------------------------------------------------------------
# Client
# ---------------------------------------------------------------------------

class BitcoinClient:
    """JSON-RPC client for the Bitcoin node."""

    def __init__(self):
        self._endpoint = os.getenv("BTC_RPC_ENDPOINT")
        self._user     = os.getenv("BTC_RPC_USER")
        self._password = os.getenv("BTC_RPC_PASS")
        self._timeout  = 30

        if not self._endpoint:
            raise EnvironmentError("Missing BTC_RPC_ENDPOINT in .env")
        if not self._user:
            raise EnvironmentError("Missing BTC_RPC_USER in .env")
        if not self._password:
            raise EnvironmentError("Missing BTC_RPC_PASS in .env")

        self._session = requests.Session()
        self._session.auth = HTTPBasicAuth(self._user, self._password)
        self._session.headers.update({"Content-Type": "application/json"})

    def call(self, method: str, params: list = None) -> dict:
        """Send a JSON-RPC 1.0 request to the Bitcoin node.

        Args:
            method: Bitcoin RPC method name (e.g. "getblockchaininfo").
            params: List of parameters. Default: [].

        Returns:
            The result field of the JSON-RPC response.

        Raises:
            RuntimeError: On RPC-level errors returned by the node.
            Exception:    On connection or HTTP errors.
        """
        if params is None:
            params = []

        payload = {
            "jsonrpc": "1.0",
            "id":      "multichain-explorer",
            "method":  method,
            "params":  params,
        }

        response = self._session.post(
            self._endpoint,
            json=payload,
            timeout=self._timeout,
        )
        response.raise_for_status()

        data = response.json()
        if data.get("error") is not None:
            raise RuntimeError(data["error"])

        return data["result"]

File: server/frontend/src/scripts/bitcoin_client.js
/**
 * server/frontend/src/scripts/bitcoin_client.js
 * ==============================================
 * Browser-side client for the Bitcoin RPC proxy.
 *
 * Sends requests to the Flask backend via Nginx at /bitcoin.
 * The session cookie is attached automatically by the browser.
 * No credentials needed on the frontend.
 *
 * Usage:
 *   import BitcoinRPC from "./bitcoin_client.js";
 *
 *   const rpc = new BitcoinRPC();
 *   const info = await rpc.getblockchaininfo();
 */

class BitcoinRPC {
    /**
     * @param {number} timeout - Request timeout in milliseconds. Default: 30000.
     */
    constructor(timeout = 30000) {
        this._url     = "/bitcoin";
        this._timeout = timeout;
        this._headers = {
            "Content-Type": "application/json",
        };
    }

    /**
     * Send a request to the Bitcoin RPC proxy.
     *
     * @param {string} method    - Bitcoin RPC method name.
     * @param {Array}  params    - Positional parameters. Default: [].
     * @param {string} requestId - Arbitrary ID echoed back by the node.
     * @returns {Promise<*>}     - The result field of the JSON-RPC response.
     * @throws {Error}           - On HTTP errors or RPC-level errors.
     */
    async call(method, params = [], requestId = "multichain-explorer") {
        const payload = {
            method,
            params,
            id: requestId,
        };

        const controller = new AbortController();
        const timer = setTimeout(() => controller.abort(), this._timeout);

        let response;
        try {
            response = await fetch(this._url, {
                method:      "POST",
                headers:     this._headers,
                body:        JSON.stringify(payload),
                signal:      controller.signal,
                credentials: "same-origin",
            });
        } finally {
            clearTimeout(timer);
        }

        if (!response.ok) {
            throw new Error(`HTTP error: ${response.status} ${response.statusText}`);
        }

        const data = await response.json();

        if (data.error != null) {
            throw new Error(JSON.stringify(data.error));
        }

        return data.result;
    }

    // ------------------------------------------------------------------
    // Convenience shortcuts
    // ------------------------------------------------------------------

    async getblockchaininfo() { return this.call("getblockchaininfo"); }
    async getdeploymentinfo()  { return this.call("getdeploymentinfo"); }
    async verifychain()        { return this.call("verifychain"); }
    async getchainstates()     { return this.call("getchainstates"); }
    async getchaintips()       { return this.call("getchaintips"); }
    async getdifficulty()      { return this.call("getdifficulty"); }
    async getmempoolinfo()     { return this.call("getmempoolinfo"); }
    async getrawmempool()      { return this.call("getrawmempool"); }
    async getblockcount()      { return this.call("getblockcount"); }
    async getbestblockhash()   { return this.call("getbestblockhash"); }

    /**
     * Returns all built-in shortcut methods as callable entries.
     * Useful for dynamic dispatch in the dashboard.
     */
    getMethods() {
        return {
            getblockchaininfo: () => this.getblockchaininfo(),
            getdeploymentinfo: () => this.getdeploymentinfo(),
            verifychain:       () => this.verifychain(),
            getchainstates:    () => this.getchainstates(),
            getchaintips:      () => this.getchaintips(),
            getdifficulty:     () => this.getdifficulty(),
            getmempoolinfo:    () => this.getmempoolinfo(),
            getrawmempool:     () => this.getrawmempool(),
            getblockcount:     () => this.getblockcount(),
            getbestblockhash:  () => this.getbestblockhash(),
        };
    }
}

export default BitcoinRPC;





